### PR TITLE
Clean up our testing

### PR DESCRIPTION
**Pull request by ebsmothers**
### Context
Our tests are a mess. This PR is an attempt to make things better.

**What would a good testing setup look like?**

There are a couple things we need to consider: (1) we have recipes that are not part of our core package and we still want to test them. (2) A lot of our recipes are distributed, and we need to test them on 2+ GPUs. We also need to set things up so that (a) we are reasonably confident that our test suite will catch breakages on bad PRs, and (b) we do not waste a ton of dev time (waiting for CI) and compute to accomplish that. 

So I propose the following delineation:

1) Unit tests. This is what's in `tests/torchtune` currently. Tbh I haven't done much with these in this PR, mainly because they're in decent (enough) shape already.
2) ~~Integration~~ Recipe tests. These should directly call the CLI. They should use the prod config, but *deliberately* override (a) the model (we should not use the full 7B) and (b) other configs as needed within the test class (e.g. setting `max_steps_per_epoch=2`). Recipe test files should maintain a 1:1 mapping with recipes.
3) Regression tests. In this PR it's still pretty minimal, but these are where we will actually run decently-sized fine-tunes and/or evals of full-size models and check the performance. These should only run on a nightly cadence, no need to block everyone's PRs with these heavyweight tests.

**We can now also handily mark and run each of these categories of tests** (as originally suggested in #296). Just add `@pytest.mark.integration_test` to your integration tests, and `@pytest.mark.slow_integration_test` to your regression tests. This means we can easily run e.g.

- Only unit tests: `pytest tests --without-integration --without-slow-integration`
- Only recipe tests: `pytest tests -m integration_test`
- Only regression tests: `pytest tests -m slow_integration_test`

(or really any other subset of these 3 you want). 

Other stuff: 
- (My fav topic) When testing a new component **DO NOT USE init_weights_with_constant**. Use `fixed_init_model` instead. Did you know that `A*x != x*A`? If you use `init_weights_with_constant` you will in fact not know that.
- There are a lot of grey areas between unit tests and recipe tests. As much as possible, please try to test logic in unit tests. The current setup is that unit tests are single-device only, so anything that may behave differently on multiple devices (e.g. stuff with FSDP `FULL_SHARD` shard strategy). That being said, there are still ways to mock basic distributed functionality in unit tests (see e.g. [single_box_init](https://github.com/pytorch/torchtune/blob/240cff850c637b5bd02b6ba80dc0957a6981bd55/tests/test_utils.py#L97) or uses of [get_pet_launch_config](https://github.com/pytorch/torchtune/blob/240cff850c637b5bd02b6ba80dc0957a6981bd55/tests/conftest.py#L15)).
- If you do not add a test for a new piece of functionality you're adding or (god forbid) you remove an existing test because "it's annoying/hard to understand" I will yell at you. If it's hard to understand please make it better, I promise on a case-by-case basis it is really not that hard.

### Changelog
- Set up workflow for distributed testing of recipes on 2x GPUs
- Separate out our recipe tests into single device and distributed to match the new structure of our recipes.
- Get our recipe tests to _actually_ run properly in a distributed setting. No more monkey patching, no more reading from stdout, no more `single_box_init`, etc. etc. There is probably more here but this is at least a minimal set of things we absolutely need to change.
- Add `gpu_test` decorator to define tests than should only run when >= n GPUs are available.
- Add back a bunch of test cases that got deleted. There are still areas missing coverage, but we are at least no worse here than we were previously.
- Say it with me: "No more separate test configs". These hurt my brain and are a maintenance nightmare. Just define overrides for each test class as needed via a list of strings. 
- Get rid of our "integration tests" altogether (along with corresponding workflow). This is just a parity check with HF. Tbh I have half a mind to delete our generation test as well, but haven't gotten there yet.
- Explicitly install pytorch in our workflow files. See #505 for more info on this. I know [this comment](https://github.com/pytorch/torchtune/issues/341#issuecomment-1963951352) mentions we should only test against nightlies, tbh I am still not 100% sold on it, but left it out of this first version for convenience more than anything.
- Remove 7B tests from `tests/recipes`. This allows us to scrap all the `--large-scale` args that we have littered everywhere (one exception is the checkpoint test, which can be done in a follow-up).
- Add two initial regression tests. The first just checks a few loss values for our llama2_7b full finetune with FSDP on 2 devices (similar to what we had previously in our single device test). The second runs LoRA FT on 7B for 500 steps, then 100 steps of eval on truthfulqa_mc2 and confirms that the accuracy is above 0.4. This is just a first pass, as we get more rigorous we can consider tightening the constraints and adding more here.
- Autodownload assets as needed. 
- Add tags for recipe tests and regression tests (discussed above).
- Combine the above two so that you can run e.g. `pytest tests --without-slow-integration` and pytest will download only small model files plus tokenizer (assuming they aren't already available locally). 

### Test plan

```
pytest tests
...
============================= 194 passed, 9 warnings in 676.25s (0:11:16) ========================
```

**Note:** Due to some challenges with the distributed CI I am gonna scrap the regression test github workflow for the time being. But the tests run and pass when run on our machines. We can make adding the CI job a fast follow.

CI is green (finally). And all I had to do was remove that pesky regression test job..

Screenshot 2024-03-22 at 5 15 13 PM

### CODE REVIEW

File: tests/assets/tiny_llama2_checkpoint.pt
Comment (joecummings) [inline tests/assets/tiny_llama2_checkpoint.pt:1]: Why delete?

Comment (ebsmothers) [reply tests/assets/tiny_llama2_checkpoint.pt:1]: Are we using it anywhere?

File: tests/regression_tests/test_llama2_7b.py
Comment (ebsmothers) [inline tests/regression_tests/test_llama2_7b.py:97]: TODO: change before landing

File: tests/regression_tests/test_llama2_7b.py
Comment (ebsmothers) [inline tests/regression_tests/test_llama2_7b.py:110]: TODO: change before landing

File: .github/workflows/regression_test.yaml
Comment (ebsmothers) [inline .github/workflows/regression_test.yaml:7]: TODO: update before landing

File: .github/workflows/recipe_test_multi_gpu.yaml
Comment (ebsmothers) [reply .github/workflows/recipe_test_multi_gpu.yaml:55]: Yeah update: I have removed large-scale stuff from our recipe tests altogether and put into regression tests.

File: tests/assets/tiny_llama2_checkpoint.pt
Comment (joecummings) [reply tests/assets/tiny_llama2_checkpoint.pt:1]: I mean we were using it in the recipe tests...

File: tests/recipes/conftest.py
Comment (rohan-varma) [reply tests/recipes/conftest.py:13]: Curious, why? just so i iunderstand

File: tests/assets/tiny_llama2_checkpoint.pt
Comment (ebsmothers) [reply tests/assets/tiny_llama2_checkpoint.pt:1]: We aren't currently on [main](https://github.com/search?q=repo%3Apytorch%2Ftorchtune%20tiny_&type=code) though, right?

File: tests/recipes/test_alpaca_generate.py
Comment (rohan-varma) [inline tests/recipes/test_alpaca_generate.py:32]: Not for this PR, but we should probably have this recipe hook into our config system + be generally better supported. Or, just rip it out entirely. cc @RdoubleA @joecummings

File: tests/recipes/test_full_finetune_single_device.py
Comment (rohan-varma) [inline tests/recipes/test_full_finetune_single_device.py:171]: What's the runtime of this test? Is there value on testing on multiple pairs of full, micro batch size?

File: tests/recipes/test_lora_finetune_distributed.py
Comment (rohan-varma) [inline tests/recipes/test_lora_finetune_distributed.py:221]: So overall I think the organization of these tests are currently fine (LoRA, full, and distributed in different recipes), but what do you think about the ability for these tests to scale? Wonder if we should eventually just generalize these to be more recipe agnostic than what they currently are.

File: tests/regression_tests/test_llama2_7b.py
Comment (rohan-varma) [inline tests/regression_tests/test_llama2_7b.py:28]: Nice! Any reason we do this only for LoRA?

**Review by rohan-varma — APPROVED** [review]
Super awesome, so excited to have our tests in a much better state! Mostly small comments and a few requiring discussion, thanks!

File: .github/workflows/conversion_script_test.yaml
Comment (ebsmothers) [reply .github/workflows/conversion_script_test.yaml:65]: Yeah this part I have not yet taken a stance on here. In practice, we can just add different versions to the matrix. In my experience running CI on nightlies means it will be red decently often. If we are OK with that I am happy to do it, but personally wanna discuss with folks a bit more on the pros and cons here.

File: .github/workflows/regression_test.yaml
Comment (ebsmothers) [reply .github/workflows/regression_test.yaml:52]: Honestly no, this is a good point. We can add a check for this, though not sure what the best way to do it is tbh. Might need a bit more research here

File: tests/cache_assets.sh
Comment (ebsmothers) [reply tests/cache_assets.sh:5]: Crap this should've been removed. Naming is hard and I couldn't make up my mind

File: tests/recipes/test_alpaca_generate.py
Comment (ebsmothers) [reply tests/recipes/test_alpaca_generate.py:32]: My vote is for the latter 😃

File: tests/recipes/test_full_finetune_single_device.py
Comment (ebsmothers) [reply tests/recipes/test_full_finetune_single_device.py:171]: Just ran this one and it actually only took about 5 seconds (probably because the model is small and we're on single device). Re multiple pairs it is kind of a holdover from the original version, and actually there was a point where the `(2, 1)` case was passing as a false positive (negative?) due to too high of a tolerance. But now that we've cleaned it up maybe we can remove it? Honestly given that the test is relatively quick I don't have a strong preference here either way.

File: tests/regression_tests/test_llama2_7b.py
Comment (ebsmothers) [reply tests/regression_tests/test_llama2_7b.py:28]: Ugh this is the one piece I have not ironed out yet. I had the full finetune test_loss in here on 7B but [took it out](https://github.com/pytorch/torchtune/pull/537/commits/b60e9d41eab0062a35e44a791ea8818453454227) just for testing. Basically all our regression tests are OOMing on the CI hardware, so I am trying to sort that out. Once it's figured out I will add the full finetune loss test back (but probably not an eval test, given that the eval results on many full finetunes are questionable anyways)

File: tests/recipes/test_lora_finetune_distributed.py
Comment (ebsmothers) [reply tests/recipes/test_lora_finetune_distributed.py:221]: Yeah I can see that being valuable down the line. The main question is how our recipes scale, depending on that it may be easier or harder to generalize our recipe tests. But in any scenario I would still like to maintain at least some version of a 1:1 mapping between recipes and recipe tests (just maybe the tests can eventually come from more of a factory pattern or something).

File: .github/workflows/regression_test.yaml
Comment (joecummings) [reply .github/workflows/regression_test.yaml:52]: Also, we shouldn't need to explicitly install the requirements. Instead we should use `pip install -e ".[dev]"`

File: .github/workflows/conversion_script_test.yaml
Comment (kartikayk) [reply .github/workflows/conversion_script_test.yaml:65]: I guess it would depend on whether we gate on any nightly features? I think ultimately we won't have an option, but would love to see if we can hold off for as long as possible. The only caveat to this is that we also want to catch pushes to torch which break us.

File: .github/workflows/recipe_test.yaml
Comment (kartikayk) [inline .github/workflows/recipe_test.yaml:48]: Curious if there's a reason to pin to this version? If so, maybe a comment to make sure no one inadvertently changes this?

File: .github/workflows/recipe_test.yaml
Comment (kartikayk) [inline .github/workflows/recipe_test.yaml:51]: this is an extreme nit, but I see a few ```python3 pip installs``` and some ```python pip installs```. Let's consolidate unless there's a reason not to?

File: tests/cache_artifacts.sh
Comment (kartikayk) [inline tests/cache_artifacts.sh:13]: YES!!!!

File: tests/test_utils.py
Comment (kartikayk) [inline tests/test_utils.py:178]: I've come to appreciate how useful pathlib is for all of this. I think you can simplify this a lot of if you convert tmpdir into a ```Path``` object (I forgot if it's already a ```Path```). In that cass this begins:

```
filename = Path.joinpath(tmpdir, suffix).with_suffix('txt')
```

File: tests/recipes/test_full_finetune_single_device.py
Comment (kartikayk) [reply tests/recipes/test_full_finetune_single_device.py:171]: If we feel confident, I'd be up for chipping away at the total running time by removing some tests

**Review by kartikayk — APPROVED** [review]
This is such an AWESOME PR! Thanks so much for fixing our tests. Conceptually this framework makes a lot of sense as well. Just a few nits. But this looks great!

File: .github/workflows/recipe_test.yaml
Comment (ebsmothers) [reply .github/workflows/recipe_test.yaml:48]: Honestly I have no idea, I am just following along with what we had previously here. Can take a look though

File: tests/recipes/test_full_finetune_single_device.py
Comment (ebsmothers) [reply tests/recipes/test_full_finetune_single_device.py:171]: Sounds good, will remove the (2, 1) case then.

File: tests/test_utils.py
Comment (ebsmothers) [reply tests/test_utils.py:178]: Sorry, actually gonna leave this as-is. The point here is to generate a unique and deterministic filename per test case here. So I am actually deliberately taking tmpdir and removing all the `/`s to do this. In that case I think Path actually makes things look a bit more complicated, since I am doing some pretty non-Pathy operations

File: tests/regression_tests/test_llama2_7b.py
Comment (ebsmothers) [reply tests/regression_tests/test_llama2_7b.py:111]: Leaving as-is as we are not running this as a nightly CI job yet anyways

File: tests/regression_tests/test_llama2_7b.py
Comment (ebsmothers) [reply tests/regression_tests/test_llama2_7b.py:124]: Same comment as [above](https://github.com/pytorch/torchtune/pull/537/files#r1536429488)

**Pull request merged**
